### PR TITLE
Jb/glns

### PR DESCRIPTION
**Pull request by j-bernardi**
caching changes to glns

### CODE REVIEW

File: dist_q_learning/q_estimators.py [hunk 1]
```diff
@@ -584,232 +596,177 @@ def update(self, history_batch):
...
+                # TODO - incorporate gamma or keep approx?
+                q_targets = IV_is / h + future_qs * (h - 1) / h
+            if debug:
+                print("Doing update using IRE uncertainty...")
+                print(f"s=\n{states}")
+                print(f"action=\n{update_action}")
+                print(f"IRE q_targets combined=\n{q_targets}")
+            # TODO lr must be scalar... Also what?
+            # TODO - do this update at the end, rather than use the model
+            #  for the next transition estimate?
```
Comment (peterbarnettnz) [inline dist_q_learning/q_estimators.py:683]: I think it's correct to do the update now, then do the transition update...? At least thats what we did for the grid world, and is what is written in the fabled Algorithm 3

File: dist_q_learning/q_estimators.py [hunk 2]
```diff
@@ -584,232 +596,177 @@ def update(self, history_batch):
...
-                # TODO - batch the learning rate!
-                if states.shape[0] <= 10:
+                if debug:
+                    print(f"Scaled:\n{q_target_transitions}")
                     print("Learning scaled transition Qs")
-                self.update_estimator(
-                    states[idxs], actions[idxs], q_target_transitions,
-                    lr=self.get_lr(ns=trans_ns),
-                    horizon=None if self.horizon_type == "inf" else h)
+            # TODO - batch the learning rate?
```
Comment (peterbarnettnz) [inline dist_q_learning/q_estimators.py:731]: This looks right to me on first glance. I am a bit unsure about the which learning rate, but what it is currently seems sensible

File: dist_q_learning/q_estimators.py [hunk 3]
```diff
@@ -825,17 +782,16 @@ def get_lr(self, ns=None):
```
Comment (peterbarnettnz) [inline dist_q_learning/q_estimators.py:782]: When `ns` is super low (close to 1) then this will mean we have a big learning rate, is this ok?

**Review by peterbarnettnz — COMMENTED** [review]
Reviewed all the substantial files (I think). Everything looks super good!! I think I understand things a lot more now. I left a few comments where I was unsure about things

File: dist_q_learning/agents.py [hunk 4]
```diff
@@ -1296,86 +1435,105 @@ def make_estimators(self):
...
             agent_value_too_low = values[proposed_action] <= scaled_min_r
+            if self.debug_mode:
+                print(f"Agent value={values[proposed_action]:.4f}")
+                print(f"Mentor value={mentor_value[0]:.4f} - "
+                      f"eps={eps_val:.4f} "
+                      + ("not scaled " if not self.scale_q_value else " ") +
+                      f"query={agent_value_too_low or prefer_mentor}")
             if agent_value_too_low or prefer_mentor:
-                action = self.mentor(state)
+                action = self.mentor(state, invert=self.invert_mentor)
```
[the remark this answers is not in the document]
Comment (j-bernardi) [reply dist_q_learning/agents.py:1489]: ah it's because I added a mentor target of moving left to right (target_centre=False). Invert mentor means it moves right to left

File: dist_q_learning/main.py [hunk 5]
```diff
@@ -353,21 +388,16 @@ def selected_mentor(state, kwargs=None):
...
 
-    lr = args.learning_rate if args.learning_rate is not None else (
-        1. if str(args.trans) == "2" else 0.1)  # 1. if deterministic
-
     if args.n_steps > 0:
         agent = agent_init(
             num_actions=env.num_actions,
             env=env,
-            gamma=0.99,
+            gamma=0.95,
```
[the remark this answers is not in the document]
Comment (j-bernardi) [reply dist_q_learning/main.py:395]: I think it helped at one point - perhaps allows immediate reward to pump up rewards faster, because cartpole doesn't need to see *too* far into the future, only far enough to see if this trajectory makes it fall over

File: dist_q_learning/mentors.py [hunk 6]
```diff
@@ -179,29 +185,53 @@ def cartpole_safe_mentor(state, kwargs=None):
...
+    Kwargs:
+        centre_coord (float): coordinate of the central x-coordinate
+        invert (Optional[bool]): used with target_centre=False, whether
+            target velocity is positive (False) or negative (True)
+
+    Kwargs:
+        centre_coord (float): the central co-ordinate. Required.
+    """
+    centre_coord = kwargs["centre_coord"]  # required
+    invert = kwargs.get("invert", False)
```
[the remark this answers is not in the document]
Comment (j-bernardi) [reply dist_q_learning/mentors.py:209]: This is used with a setting in the mentor that has a target velocity, rather than target position

**Review by j-bernardi — COMMENTED** [review]
sorry I was slow getting to it. Addressed 1 comment, and hopefuyl answered about the mentor option that targets a constant positive or (inverted) negative velocity, rather than targeting the centre

File: dist_q_learning/agents.py [hunk 7]
```diff
@@ -1296,86 +1435,105 @@ def make_estimators(self):
...
             scaled_min_r = self.min_reward
-            eps = self.epsilon()
             if not self.scale_q_value:
                 scaled_min_r /= (1. - self.gamma)
-                eps /= (1. - self.gamma)
-            mentor_value = self.mentor_q_estimator.estimate(
-                np.expand_dims(state, 0))
-            if self.batch_size <= 10:
-                print("mentor value", mentor_value)
+                mentor_pref_magnitude *= (1. - self.gamma)  # scale down
```
[the remark this answers is not in the document]
Comment (j-bernardi) [reply dist_q_learning/agents.py:1477]: Yes - epsilon will always be between 0 and 1 so we scale down the magnitude just for that

File: dist_q_learning/estimators.py [hunk 8]
```diff
@@ -478,92 +482,28 @@ def estimate(self, states, estimate_model=None):
...
         """Algorithm 1. Use experience to update estimate of immediate r
 
         Args:
-            history_batch (list[tuple]): list of (state, reward) tuples
-                that will form the batch.
+            history_batch (tuple[jnp.ndarray]|list[tuple]): list of
+                (state, reward) tuples that will form the batch. Or the
+                (states, rewards) tuple.
             update_model: the model to perform the update on.
+            tup: determines which format history comes in (TEMP)
```
[the remark this answers is not in the document]
Comment (j-bernardi) [reply dist_q_learning/estimators.py:493]: Yeah I guess it's not temporary anymore - removed

File: dist_q_learning/estimators.py [hunk 9]
```diff
@@ -419,6 +421,7 @@ def __init__(
                 for the GGLN
             context_dim (int): the number of hyperplanes used to make the 
                 halfspaces
+            feat_mean (float):
```
[the remark this answers is not in the document]
Comment (j-bernardi) [reply dist_q_learning/estimators.py:424]: yeah, so 0 or 1. Will add comment, good point

File: dist_q_learning/q_estimators.py (hunk 3 above)
Comment (j-bernardi) [reply dist_q_learning/q_estimators.py:782]: I think self.lr has never been None - I am not sure if this is okay, but I think it's what we did previously for that "take the mean" thing?

**Comment by j-bernardi** [comment]
Merged - been hanging around long enough and I didn't see any dissent in the review, only queries
